README.md
# 2HDECAY

A program for the Calculation of Electroweak One-Loop Corrections to Higgs Decays in the Two-Higgs-Doublet Model Including State-of-the-Art QCD Corrections

## Program information

**Program** 2HDECAY 1.1.3

**Date** 03.06.2019

**Authors** [Marcel Krause](mailto:marcel.krause@kit.edu), [M. Margarete Mühlleitner](mailto:margarete.muehlleitner@kit.edu) and [Michael Spira](mailto:michael.spira@psi.ch)

**Manual** https://arxiv.org/abs/1810.00768

**Citations** When you use this program, please acknowledge the work of our and other groups by citing the following papers:
- The manual for 2HDECAY:
  - M. Krause, M. Muhlleitner, M. Spira, arXiv:1810.00768
- The manuals for HDECAY:
  - A. Djouadi, J. Kalinowski, M. Spira, Comp. Phys. Commun. 108 (1998) 56, hep-ph/9704448
  - A. Djouadi, J. Kalinowski, M. Muhlleitner, M. Spira, arXiv:1801.09506 (hep-ph)
- The papers on the electroweak correction to the 2HDM decays:
  - M. Krause, R. Lorenz, M. Muhlleitner, R. Santos, H. Ziesche, JHEP 1609 (2016) 143, arXiv:1605.04853 (hep-ph)
  - M. Krause, M. Muhlleitner, R. Santos, H. Ziesche, Phys.Rev. D95 (2017) no.7, 075019, arXiv:1609.04185 (hep-ph)
  - A. Denner, S. Dittmaier, J.-N. Lang, J. High Energ. Phys. (2018) 2018: 104, arXiv:1808.03466 (hep-ph)
- The publication of LoopTools:
  - T. Hahn, M. Perez-Victoria, Comp. Phys. Commun. 118 (1999) 153-165, hep-ph/9807565

**Abstract** We present the program package 2HDECAY for the calculation of the partial decay widths and branching ratios of the Higgs bosons of a general CP-conserving 2-Higgs doublet model (2HDM). The tool includes the full electroweak one-loop corrections to all two-body onshell Higgs decays in the 2HDM that are not loop-induced. It combines them with the state-of-the-art QCD corrections that are already implemented in the program HDECAY. For the renormalization of the electroweak sector an on-shell scheme is implemented for most of the renormalization parameters. Exceptions are the soft-![](https://latex.codecogs.com/gif.latex?%5Cmathbb%7BZ%7D_2 "\mathbb{Z}_2")-breaking squared mass scale ![](https://latex.codecogs.com/gif.latex?m_%7B12%7D%5E2 "m_{12}^2"), where an ![](https://latex.codecogs.com/gif.latex?%5Coverline%7B%5Ctext%7BMS%7D%7D "\overline{\text{MS}}") condition is applied, as well as the 2HDM mixing angles ![](https://latex.codecogs.com/gif.latex?%5Calpha "\alpha") and ![](https://latex.codecogs.com/gif.latex?%5Cbeta "\beta"), for which several distinct renormalization schemes are implemented. The tool 2HDECAY can be used for phenomenological analyses of the branching ratios of Higgs decays in the 2HDM. Furthermore, the separate output of the electroweak contributions to the tree-level partial decay widths for several different renormalization schemes allows for an efficient analysis of the impact of the electroweak corrections and the remaining theoretical error due to missing higher-order corrections. The latest version of the program package 2HDECAY can be downloaded from the URL https://github.com/marcel-krause/2HDECAY.

**Changelog** For a documentation about the changes made in 2HDECAY, check the [Changelog.md](Changelog.md) file.

**Copyright** Copyright (C) 2018-2019, Marcel Krause, Milada Margarete Mühlleitner and Michael Spira

**License** GNU General Public License (GNU GPL-3.0-or-later). 2HDECAY is released under GNU General Public License (GNU GPL-3.0-or-later). This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details. You have received a copy ([LICENSE.md](LICENSE.md)) of the GNU General Public License along with this program.

**Contact** For feedback, complaints and bug reports, please send an e-mail to <marcel.krause@kit.edu>, <margarete.muehlleitner@kit.edu> and <michael.spira@psi.ch>.

## Getting Started

For more detailed information about installing and using the program, please have a look at the [full manual](Documentation/manual.pdf "full manual").

### System requirements

Supported operating systems:
- Windows 7 and Windows 10 (with [Cygwin](https://www.cygwin.com/ "Cygwin") installed)
- Linux
- macOS

The following components must be installed on your system to compile and run 2HDECAY:
- GNU gcc (tested with versions 6.4.0 and 7.3.1)
- GNU g++
- GNU gfortran
- Python 2 or 3 (tested with versions 2.7.14 and 3.5.0)
- find
- cURL

If you install 2HDECAY on Windows, make sure that you install [Cygwin](https://www.cygwin.com/ "Cygwin") first, including the following components:
- GNU gcc
- GNU g++
- GNU gfortran
- find
- cURL

### Installing

For an easy installation, we recommend using the automatic installer of 2HDECAY, which guides you through the installation. Download the latest version of 2HDECAY from https://github.com/marcel-krause/2HDECAY. Open a shell, navigate to the 2HDECAY root folder and execute the following command:
```
python setup.py
```

The installer asks you if you want to download and install LoopTools, which is required for 2HDECAY to run. If you do not have LoopTools installed on your system already type y. The installer downloads the LoopTools version specified in the file [Config.py](Config.py); if you prefer another version, change the corresponding entry in that file first before executing the setup.
If you already have LoopTools on your system, you can choose not to download and install LoopTools. In that case, open the file [Config.py](Config.py) and set
```
useRelativeLoopToolsPath = False
```
and change the variable pathLoopTools to the *absolute* path to the LoopTools main folder (containing the bin, include and library subfolders). Additionally, modify the variables pathLoopToolsLibs and pathLoopToolsExecs to the relative paths to the library and bin subfolders with respect to the LoopTools main directory.

The installer further asks you whether the makefile and electroweakCorrections.F90 file shall be created and whether it should make the program. Type y for all these questions in order to compile 2HDECAY.

### Using 2HDECAY

Using 2HDECAY is simple. Save all input files that you want to use in the 'Input' subdirectory. You can put as many input files in the directory as you wish and the input files can have arbitrary file names. Make sure that the input files have the exact input format required by 2HDECAY, as described in the [full manual](Documentation/manual.pdf "full manual").

To start 2HDECAY, execute the following command in the main folder of 2HDECAY:
```
python 2HDECAY.py
```
The program will iterate over all input files and calculate all higher-order corrections according to the parameter and renormalization scheme choices as given in your input files. The resulting output files are saved with the same file name as the input files in the 'Results' subdirectory. For each input file, two output files are generated: one file containing the branching ratios with and without the electroweak corrections, indicated by a filename suffix '_BR', as well as one file containing the electroweak partial decay widths at tree-level and one-loop order, indicated by the filename suffix '_EW'.

Changelog.md
## [1.1.3] - 2019-06-03
### Changed
- Fixed a bug which prevented the program to compute the electroweak partial decay widths if RENSCHEM=0 was set.

## [1.1.2] - 2019-05-08
### Changed
- Removed the gitignore statement for the makefile in the subfolder HDECAY in order to track changes made in this file.

## [1.1.1] - 2019-04-03
### Changed
- Fixed some minor bugs in the rescaling of the Fermi constant for some decays in the HDECAY subroutine (the numerical effect of the bugs on the branching ratios and partial decay widths is negligible).

## [1.1.0] - 2019-02-22
### New
- Implemented the parameter conversion for directly comparing results computed in distinct renormalization schemes.
- Implemented the scale evolution of ![](https://latex.codecogs.com/gif.latex?%5Coverline%7B%5Ctext%7BMS%7D%7D "\overline{\text{MS}}") input parameters from an input renormalization scale (INSCALE) to the renormalization scale at which the NLO calculations are performed (OUTSCALE).
- Implemented six new renormalization schemes for the scalar mixing angles ![](https://latex.codecogs.com/gif.latex?%5Calpha "\alpha") and ![](https://latex.codecogs.com/gif.latex?%5Cbeta "\beta"):
  - "OS1 scheme" as described in *A. Denner, S. Dittmaier, J.-N. Lang, J. High Energ. Phys. (2018) 2018: 104* with scheme identifier 12.
  - "OS2 scheme" as described in *A. Denner, S. Dittmaier, J.-N. Lang, J. High Energ. Phys. (2018) 2018: 104* with scheme identifier 13.
  - "OS12 scheme" as described in *A. Denner, S. Dittmaier, J.-N. Lang, J. High Energ. Phys. (2018) 2018: 104* with scheme identifier 14.
  - "BFMS scheme" as described in *A. Denner, S. Dittmaier, J.-N. Lang, J. High Energ. Phys. (2018) 2018: 104* with scheme identifier 15.
  - ![](https://latex.codecogs.com/gif.latex?%5Coverline%7B%5Ctext%7BMS%7D%7D "\overline{\text{MS}}") (standard tadpole scheme) with scheme identifier 16.
  - ![](https://latex.codecogs.com/gif.latex?%5Coverline%7B%5Ctext%7BMS%7D%7D "\overline{\text{MS}}") (alternative tadpole scheme) with scheme identifier 17.
### Removed
- Removed the three process-dependent schemes "proc.-dep. 1 (standard tadpole scheme)", "proc.-dep. 2 (standard tadpole scheme)" and "proc.-dep. 3 (standard tadpole scheme)" due to redundancy, since they yield the same result as the process-dependent schemes defined in the alternative tadpole scheme.
### Changed
- Updated the manual according to all changes made in version 1.1.0.
- The input file format was changed. Trying to use the input file format that was used in 2HDECAY 1.0.0, 1.0.1 and 1.0.2 will throw an error and exceptionally terminate the program.. The following lines changed in the new input file format:
  - in line 59 of the input file, an additional input parameter REFSCHEM was added which is used for defining the reference scheme for the parameter conversion. 
  - in line 64 of the input file, an additional input parameter OUTSCALE was added which overtakes the function of the formerly defined INSCALE, i.e. OUTSCALE is the renormalization scale at which the one-loop integrals and counterterms are computed.
- The input parameter INSCALE, defined in line 63 of the input file (according to the new input file format), now has a new meaning compared to 2HDECAY 1.0.0, 1.0.1 and 1.0.2. It is now the renormalization scale at which all ![](https://latex.codecogs.com/gif.latex?%5Coverline%7B%5Ctext%7BMS%7D%7D "\overline{\text{MS}}") input parameters are defined.
- The output file format was changed to account for the parameter conversion: the scalar mixing angles alpha and beta, the soft-![](https://latex.codecogs.com/gif.latex?%5Cmathbb%7BZ%7D_2 "\mathbb{Z}_2")-breaking squared mass scale ![](https://latex.codecogs.com/gif.latex?m_%7B12%7D%5E2 "m_{12}^2") and the LO decay width and branching ratios are given for each renormalization scheme that is chosen.
- Due to the removal of the three process-dependent schemes in the standard tadpole scheme, the numbering of the schemes changed:
  - The scheme "proc.-dep. 1 (alternative tadpole scheme)" now has scheme identifier 9 (former identifier: 10).
  - The scheme "proc.-dep. 2 (alternative tadpole scheme)" now has scheme identifier 10 (former identifier: 12). 
  - The scheme "proc.-dep. 3 (alternative tadpole scheme)" now has scheme identifier 11 (former identifier: 14).
- Fixed some bugs concerning the rescaling of the Fermi constant ![](https://latex.codecogs.com/gif.latex?G_F "G_F") in the HDECAY subroutine.
- Fixed a bug concerning the print-out of the values of alpha in the intermediate output file that is read in by HDECAY if RENSCHEM was set to values larger than 9.

## [1.0.2] - 2018-10-30
### Changed
- Fixed some typos in the Yukawa couplings for the real corrections to leptonic decays in 2HDM types X and Y.

## [1.0.1] - 2018-10-29
### Changed
- Fixed a bug that would prevent the compilation of 2HDECAY when the path to 2HDECAY would contain whitespace characters (on all operating systems).
- Updated some copyright information.

2HDECAY.py
#!/usr/bin/env python
#Filename: 2HDECAY.py


 ##########################################################################################################################################
#                                                                                                                                          #
#                                                                2HDECAY                                                                   #
#                                                                                                                                          #
#   Purpose:    A program for the calculation One-Loop Electroweak Corrections to Higgs Decays in the                                      #
#               Two-Higgs-Doublet Model (2HDM) Including State-of-the-Art QCD Corrections                                                  #
#   Authors:    Marcel Krause (marcel.krause@kit.edu)                                                                                      #
#               Prof. Dr. M. Margarete Muehlleitner (margarete.muehlleitner@kit.edu)                                                       #
#               Dr. Michael Spira (michael.spira@psi.ch)                                                                                   #
#   Version:    1.1.3                                                                                                                      #
#   Date:       03.06.2019                                                                                                                 #
#   Copyright:  Copyright (C) 2018-2019, Marcel Krause, Milada Margarete Muehlleitner and Michael Spira                                    #
#   License:    GNU General Public License (GNU GPL-3.0-or-later)                                                                          #
#                                                                                                                                          #
#               2HDECAY is released under GNU General Public License (GNU GPL-3.0-or-later).                                               #
#               This program is free software: you can redistribute it and/or modify it under the terms of the                             #
#               GNU General Public License as published by the Free Software Foundation, either version 3 of                               #
#               the License, or any later version.                                                                                         #
#                                                                                                                                          #
#               This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;                                  #
#               without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.                                  #
#               See the GNU General Public License for more details.                                                                       #
#                                                                                                                                          #
#               You have received a copy (LICENSE.md) of the GNU General Public License along with this program                            #
#               in the 2HDECAY root directory.                                                                                             #
#                                                                                                                                          #
#   Citation:   When you use this program, please acknowledge the work of our and other groups by citing the                               #
#               following papers:                                                                                                          #
#                   The manual for 2HDECAY:                                                                                                #
#                    - M. Krause, M. Muhlleitner, M. Spira, arXiv:1810.00768                                                               #
#                   The manuals for HDECAY:                                                                                                #
#                    - A. Djouadi, J. Kalinowski, M. Spira, Comp. Phys. Commun. 108 (1998) 56, hep-ph/9704448                              #
#                    - A. Djouadi, J. Kalinowski, M. Muhlleitner, M. Spira, arXiv:1801.09506 (hep-ph)                                      #
#                   The papers on the EW correction to the 2HDM decays:                                                                    #
#                    - M. Krause, R. Lorenz, M. Muhlleitner, R. Santos, H. Ziesche, JHEP 1609 (2016) 143, arXiv:1605.04853 (hep-ph)        #
#                    - M. Krause, M. Muhlleitner, R. Santos, H. Ziesche, Phys.Rev. D95 (2017) no.7, 075019, arXiv:1609.04185 (hep-ph)      #
#                    - A. Denner, S. Dittmaier, J.-N. Lang, J. High Energ. Phys. (2018) 2018: 104, arXiv:1808.03466 (hep-ph)               #
#                   The publication of LoopTools:                                                                                          #
#                    - T. Hahn, M. Perez-Victoria, Comp. Phys. Commun. 118 (1999) 153-165, hep-ph/9807565                                  #
#                                                                                                                                          #
 ##########################################################################################################################################


#------------------------------#
#         Import Modules       #
#------------------------------#
import sys
import os
from shutil import copyfile, rmtree
from math import pi, sqrt
import subprocess
# import multiprocessing
import CommonFunctions      # Provides common, often used functions for different scripts of 2HDECAY

#-------------------------#
#        Settings         #
#-------------------------#
# WARNING: do not change these settings if you do not know what they do!
lineToInsert = 124      # This is the line at which the temporary input file ends and at which we append the electroweak corrections
lineWhereAlphaAtMZ = 26 # This is the line at which in the temporary input file the fine-structure constant at the Z boson mass MZ is specified
lineWhereGFCalc = 28    # This is the line at which in the temporary input file the calculated Fermi constant GFCALC is specified
lineWhereMZ = 31        # This is the line at which in the temporary input file the Z boson mass MZ is specified
lineWhereMW = 32        # This is the line at which in the temporary input file the W boson mass MW is specified
lineWhereOSMC = 22      # This is the line at which the OS MC value has to be inserted in the temporary input file
lineWhereOSMB = 23      # This is the line at which the OS MB value has to be inserted in the temporary input file
lineWhereParamType = 56 # This is the line at which the parameter type is specified
lineWhereRefScheme = 59 # This is the line at which the reference renormalization scheme is specified

#----------------------------#
#        Main Program        #
#----------------------------#

if __name__ == "__main__":
	# Print the welcome screen
	print('''
+---------------------------------------+
|                                       |
|             2HDECAY 1.1.3             |
|                                       |
|                             /         |
|                            /          |
|                           /           |
|                      --- /            |
|      ______________/     \            |
|                    \     /            |
|                      --- \            |
|                           \           |
|                            \          |
|                             \         |
|                                       |
+---------------------------------------+

When you use this program please cite:
	The manual for 2HDECAY:
	 - M. Krause, M. Muhlleitner, M. Spira, arXiv:1810.00768
	The manuals for HDECAY:
	 - A. Djouadi, J. Kalinowski, M. Spira, Comp. Phys. Commun. 108 (1998) 56, hep-ph/9704448
	 - A. Djouadi, J. Kalinowski, M. Muhlleitner, M. Spira, arXiv:1801.09506 (hep-ph)
	The papers on the EW correction to the 2HDM decays:
	 - M. Krause, R. Lorenz, M. Muhlleitner, R. Santos, H. Ziesche, JHEP 1609 (2016) 143, arXiv:1605.04853 (hep-ph)
	 - M. Krause, M. Muhlleitner, R. Santos, H. Ziesche, Phys.Rev. D95 (2017) no.7, 075019, arXiv:1609.04185 (hep-ph)
	 - A. Denner, S. Dittmaier, J.-N. Lang, J. High Energ. Phys. (2018) 2018: 104, arXiv:1808.03466 (hep-ph)
	The publication of LoopTools:
	 - T. Hahn, M. Perez-Victoria, Comp. Phys. Commun. 118 (1999) 153-165, hep-ph/9807565

2HDECAY is released under GNU General Public License (GNU GPL-3.0-or-later). This program is free software: 
you can redistribute it and/or modify it under the terms of the GNU General Public License as published by 
the Free Software Foundation, either version 3 of the License, or any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the 
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You have received a copy (LICENSE.md) of the GNU General Public License along with this program in the 2HDECAY
root directory.

Copyright 2018-2019, Marcel Krause, Milada Margarete Muehlleitner and Michael Spira.
	''')

	# Get a list of all input files
	inputPath = "Input"
	inputFileList = os.listdir(inputPath)
	if '..' in inputFileList:
		inputFileList.remove('..')
	if '.' in inputFileList:
		inputFileList.remove('.')

	# Iterate over all input files
	for inputFileTemp in inputFileList:
		print("Calculating corrections for input file " + inputFileTemp + " ...\n")

		# Copy the input file to the HDECAY subfolder
		print("Copying input files into HDECAY folder...")
		filenameIn = "Input" + os.sep + inputFileTemp
		filenameOut = "HDECAY" + os.sep + "hdecay.in"
		# Remove any existing input and fermion masses file in HDECAY and prepare a template mixing angles file
		if os.path.isfile(filenameOut):
			os.remove(filenameOut)
		if os.path.isfile("HDECAY" + os.sep + "fermionmasses.dat"):
			os.remove("HDECAY" + os.sep + "fermionmasses.dat")
		if os.path.isfile("HDECAY" + os.sep + "alphaandbeta.dat"):
			os.remove("HDECAY" + os.sep + "alphaandbeta.dat")
		copyfile(filenameIn, filenameOut)
		with open("HDECAY" + os.sep + "alphaandbeta.dat", 'w') as fileHandler:
			fileHandler.write(' alpha = 0D0\n beta  = 0D0\n')
		print("... done.\n")

		# Let HDECAY run in minimal mode to produce the fermion mass file
		print("Starting HDECAY in minimal mode...")
		os.chdir('HDECAY')
		prompt = ['./run', '1']
		subprocess.call(prompt, stdin=None, stdout=None, stderr=None, shell=False)
		os.chdir('..')
		print("HDECAY in minimal mode terminated.\n")

		# Read the fermion masses from the fermion mass file
		filenameMasses = "HDECAY" + os.sep + "fermionmasses.dat"
		massFileLines = list(line.rstrip('\n') for line in open(filenameMasses))
		MCOSCalc = float((massFileLines[0].split('='))[1].strip())
		MBOSCalc = float((massFileLines[1].split('='))[1].strip())
		
		# Copy the file name to the HDECAY folder and truncate it at the end
		print("Copying input files into HDECAY folder...")
		filenameIn = "Input" + os.sep + inputFileTemp
		filenameOut = "HDECAY" + os.sep + "hdecay.in"
		if os.path.isfile(filenameOut):
			os.remove(filenameOut)
		fileHandler = open(filenameIn, "r")
		convertedFileHandler = []
		lineCount = 1
		for line in fileHandler:
			# If the parameter type is 2, then the reference renormalization scheme has to be set to zero
			if (lineCount == lineWhereRefScheme):
				refScheme = int((line.split())[2])
				if (refScheme == 0):
					print("\nERROR: REFSCHEM=0 is given as input, but REFSCHEM must be greater than zero.")
					print("2HDECAY will be terminated now (ERROR: REFSCHEM=0 was set).")
					sys.exit()
				if (refScheme != 0) and (paramType == 2):
					convertedFileHandler.append("REFSCHEM = 0\n")
					print("\nWARNING: REFSCHEM (line 59) is given as a non-zero value, but TYPE=2 (line 57, lambdas as input) is set.")
					print("REFSCHEM is overwritten to zero, automatic parameter conversion is deactivated!")
					lineCount += 1
					continue
			# Write the current line in an array
			convertedFileHandler.append(line)
			# Check for the parameter type
			if lineCount == lineWhereParamType:
				paramType = int((line.split())[2])
			# Pick out the values of MW, MZ and alphaAtMZ
			if lineCount == lineWhereMZ:
				massMZ = float((line.split())[2])
			if lineCount == lineWhereMW:
				massMW = float((line.split())[2])
			if lineCount == lineWhereAlphaAtMZ:
				alphaAtMZ = float((line.split())[2])
			lineCount += 1
		fileHandler.close()

		# Check whether the new file format (2HDECAY 1.1.0 and upwards) is used
		isNewFormat = False
		for line in convertedFileHandler:
			if "REFSCHEM" in line:
				isNewFormat = True
		if not isNewFormat:
			print("\nERROR: incompatible file format! The input file format was changed with 2HDECAY 1.1.0 to enable input parameter conversion. Please check your input files for compatibility with the new input file format, cf. the Changelog.md file and the manual of the latest 2HDECAY version.")
			print("2HDECAY will be terminated now (ERROR: incompatible file format for 2HDECAY >= 1.1.0).")
			sys.exit()

		# Write a copy of the file to the output folder, but add GFCALC, MCOSCALC and MBOSCALC with the calculated values
		GFcalc = pi/sqrt(2)*alphaAtMZ/(massMW**2*(1-massMW**2/massMZ**2))
		GFline = "GFCALC   = " + str(GFcalc) + "\n"
		MCOSline = "MCOSCALC = " + str(MCOSCalc) + "\n"
		MBOSline = "MBOSCALC = " + str(MBOSCalc) + "\n"
		lineCount = 1
		convertedFile = ''
		renScaleIsDynamic = '0'
		for line in convertedFileHandler:
			if "OUTSCALE" in line and "MIN" in line:
				renScaleIsDynamic = '1'
			if lineCount == lineWhereGFCalc:
				convertedFile += GFline
			elif lineCount == lineWhereOSMC:
				convertedFile += MCOSline
				convertedFile += MBOSline
				convertedFile += line
			else:
				convertedFile += line
			lineCount += 1
		fileHandler = open(filenameOut, "w")
		fileHandler.write(convertedFile)
		fileHandler.close()
		print("... done.\n")
		
		# Calculate the electroweak corrections
		print("Calculating electroweak corrections...\n")
		prompt = ['./electroweakCorrections', '0', '0', '0', '1', 'HDECAY' + os.sep + 'hdecay.in', 'hdecay.in', renScaleIsDynamic]
		subprocess.call(prompt, stdin=None, stdout=None, stderr=None, shell=False)
		print("Calculation of electroweak corrections done.\n")

		# Replace the newline character in each file with a proper newline
		print("Postprocessing temporary input file...")
		fileHandler = open(filenameOut, "r")
		convertedFile = ''
		lineCount = 1
		for line in fileHandler:
			# Convert the literal newlines to actual ones and remove the leading whitespace from the Fortran output
			if (lineCount == lineToInsert):
				lineToReplace = line.replace('\\n', '\n')[1:]
			else:
				lineToReplace = line.replace('\\n', '\n')
			convertedFile += lineToReplace
			# print(lineToReplace)
			lineCount += 1
		fileHandler.close()

		# Store the results file in the correct directory
		if os.path.isfile(filenameOut):
			os.remove(filenameOut)
		fileHandler = open(filenameOut, "w")
		fileHandler.write(convertedFile)
		fileHandler.close()
		print("... done.\n")

		# Start HDECAY in the normal (non-minimal) configuration
		print("Starting HDECAY in standard mode...")
		os.chdir('HDECAY')
		prompt = ['./run']
		subprocess.call(prompt, stdin=None, stdout=None, stderr=None, shell=False)
		os.chdir('..')
		print("HDECAY in standard mode terminated.\n")

		# Copy the output files to the results folder
		print("Copying input files into output folder...")
		filenameIn = "HDECAY" + os.sep + "slha.out"
		filenameOut = "Results" + os.sep + inputFileTemp.replace('.in', '_BR.out')
		copyfile(filenameIn, filenameOut)
		filenameIn = "HDECAY" + os.sep + "ewpartialwidth.out"
		filenameOut = "Results" + os.sep + inputFileTemp.replace('.in', '_EW.out')
		copyfile(filenameIn, filenameOut)
		print("... done.\n")

		# Cleaning 
		if os.path.isfile("HDECAY" + os.sep + "fermionmasses.dat"):
			os.remove("HDECAY" + os.sep + "fermionmasses.dat")
		if os.path.isfile("HDECAY" + os.sep + "alphaandbeta.dat"):
			os.remove("HDECAY" + os.sep + "alphaandbeta.dat")

		print("Corrections for input file " + inputFileTemp + " done.\n")

	# End of program is reached
	print("All calculations finished. Thanks for using 2HDECAY!")

	# Print the end screen 
	print('''
When you use this program please cite:
	The manual for 2HDECAY:
	 - M. Krause, M. Muhlleitner, M. Spira, arXiv:1810.00768
	The manuals for HDECAY:
	 - A. Djouadi, J. Kalinowski, M. Spira, Comp. Phys. Commun. 108 (1998) 56, hep-ph/9704448
	 - A. Djouadi, J. Kalinowski, M. Muhlleitner, M. Spira, arXiv:1801.09506 (hep-ph)
	The papers on the EW correction to the 2HDM decays:
	 - M. Krause, R. Lorenz, M. Muhlleitner, R. Santos, H. Ziesche, JHEP 1609 (2016) 143, arXiv:1605.04853 (hep-ph)
	 - M. Krause, M. Muhlleitner, R. Santos, H. Ziesche, Phys.Rev. D95 (2017) no.7, 075019, arXiv:1609.04185 (hep-ph)
	 - A. Denner, S. Dittmaier, J.-N. Lang, J. High Energ. Phys. (2018) 2018: 104, arXiv:1808.03466 (hep-ph)
	The publication of LoopTools:
	 - T. Hahn, M. Perez-Victoria, Comp. Phys. Commun. 118 (1999) 153-165, hep-ph/9807565

2HDECAY is released under GNU General Public License (GNU GPL-3.0-or-later). This program is free software: 
you can redistribute it and/or modify it under the terms of the GNU General Public License as published by 
the Free Software Foundation, either version 3 of the License, or any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the 
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You have received a copy (LICENSE.md) of the GNU General Public License along with this program in the 2HDECAY
root directory.

Copyright 2018-2019, Marcel Krause, Milada Margarete Muehlleitner and Michael Spira.
	''')

	sys.exit()
